lib/circuit.js

```diff
@@ -235,89 +238,92 @@ class CircuitBreaker extends EventEmitter {
-      /**
-       * Emitted when the circuit breaker does not find a value in
-       * the cache, but the cache option is enabled.
-       * @event CircuitBreaker#cacheMiss
-       */
-      this.emit('cacheMiss');
-    }
+    const semaphore = new Semaphore(this.options.capacity);
+    return semaphore.use(() => {
+      const args = Array.prototype.slice.call(arguments);
 
-    if (this.opened && !this.pendingClose) {
       /**
-       * Emitted when the circuit breaker is open and failing fast
-       * @event CircuitBreaker#reject
+       * Emitted when the circuit breaker action is executed
+       * @event CircuitBreaker#fire
        */
-      this.emit('reject', new Error('Breaker is open'));
+      this.emit('fire', args);
+
+      if (CACHE.get(this) !== undefined) {
+        /**
+         * Emitted when the circuit breaker is using the cache
+         * and finds a value.
+         * @event CircuitBreaker#cacheHit
+         */
+        this.emit('cacheHit');
+        return CACHE.get(this);
+      } else if (this.options.cache) {
+        /**
+         * Emitted when the circuit breaker does not find a value in
+         * the cache, but the cache option is enabled.
+         * @event CircuitBreaker#cacheMiss
+         */
+        this.emit('cacheMiss');
+      }
 
-      return fallback(this, 'Breaker is open', args, 0) ||
-        Promise.reject(new Error('Breaker is open'));
-    }
-    this[PENDING_CLOSE] = false;
+      if (this.opened && !this.pendingClose) {
+        /**
+         * Emitted when the circuit breaker is open and failing fast
+         * @event CircuitBreaker#reject
+         */
+        this.emit('reject', new Error('Breaker is open'));
 
-    let timeout;
-    let timeoutError = false;
-    return new Promise((resolve, reject) => {
-      const latencyStartTime = Date.now();
-      timeout = setTimeout(
-        () => {
-          timeoutError = true;
-          const error = new Error(`Timed out after ${this.options.timeout}ms`);
-          error.code = 'ETIMEDOUT';
-          /**
-           * Emitted when the circuit breaker action takes longer than `options.timeout`
-           * @event CircuitBreaker#timeout
-           */
-          const latency = Date.now() - latencyStartTime;
-          this.emit('timeout', error, latency);
-          resolve(handleError(error, this, timeout, args, latency, resolve, reject));
-        }, this.options.timeout);
-
-      try {
-        const result = this.action.apply(this.action, args);
-        const promise = (typeof result.then === 'function')
-          ? result
-          : Promise.resolve(result);
-
-        promise.then((result) => {
-          if (!timeoutError) {
-            clearTimeout(timeout);
+        return fallback(this, 'Breaker is open', args, 0) ||
+          Promise.reject(new Error('Breaker is open'));
+      }
+      this[PENDING_CLOSE] = false;
+
+      let timeout;
+      let timeoutError = false;
+      return new Promise((resolve, reject) => {
```

Comment:
I think we don't need to call semaphore.use() until here. All of the other stuff should be allowed to execute. E.g. I think it's safe to check for a cache hit and also to check if the circuit is open before attempting to acquire the semaphore. No reason for that stuff to be rate limited.